src/canonical/handler.js

```diff
@@ -0,0 +1,494 @@
+  // temporary, to check what input it gets
+  let baseURL = input;
+  if (!baseURL.startsWith('https://')) {
+    const site = await retrieveSiteBySiteId(dataAccess, input, log);
+    if (!site) {
+      return notFound('Site not found');
+    }
+    baseURL = site.getBaseURL();
+    log.info(`Retrieved base URL: ${baseURL} for site ID: ${input}`);
+  }
+  try {
+    const topPages = await getTopPagesForSite(baseURL, context, log);
+    log.info(`Top pages for baseURL ${baseURL}: ${JSON.stringify(topPages)}`);
+
+    if (topPages.length === 0) {
+      log.info('No top pages found, ending audit.');
+      return {
+        domain: baseURL,
+        results: [{
+          check: ChecksAndErrors.TOPPAGES.check,
+          error: ChecksAndErrors.TOPPAGES.error,
+        }],
+        success: false,
+      };
+    }
+
+    // const aggregatedPageLinks = await getBaseUrlPagesFromSitemaps(
+    //   baseURL,
+    //   topPages.map((page) => page.url),
+    // );
+    // eslint-disable-next-line max-len
+    // log.info(`Aggregated page links from sitemaps for baseURL ${baseURL}: ${JSON.stringify(aggregatedPageLinks)}`);
+
+    const auditPromises = topPages.map(async (page) => {
+      const { url } = page;
+      log.info(`Validating canonical tag for URL: ${url}`);
+      const checks = [];
+
+      const { canonicalUrl, checks: canonicalTagChecks } = await validateCanonicalTag(url, log);
+      checks.push(...canonicalTagChecks);
+
+      if (canonicalUrl) {
+        log.info(`Found canonical URL: ${canonicalUrl}`);
+        // if (canonicalUrl && !canonicalTagChecks.some((check) => check.error)) {
+        // const allPages = [];
+        // const setsOfPages = Object.values(aggregatedPageLinks);
+        // const setsOfPages = topPages;
+        // for (const pages of setsOfPages) {
+        //   allPages.push(...pages);
+        // }
+        // for (const pages of setsOfPages) {
+        //   if (Array.isArray(pages)) {
+        //     allPages.push(...pages);
+        //   } else if (pages && pages.url) {
+        //     allPages.push(pages.url);
+        //   }
+        // }
+
+        // const sitemapCheck = validateCanonicalInSitemap(allPages, canonicalUrl);
+        // checks.push(sitemapCheck);
+
+        const urlFormatChecks = validateCanonicalUrlFormat(canonicalUrl, baseURL, log);
+        log.info(`validateCanonicalUrlFormat results for ${canonicalUrl}: ${JSON.stringify(urlFormatChecks)}`);
+        checks.push(...urlFormatChecks);
+
+        const urlContentCheck = await validateCanonicalUrlContentsRecursive(canonicalUrl, log);
+        log.info(`validateCanonicalUrlContentsRecursive result for ${canonicalUrl}: ${JSON.stringify(urlContentCheck)}`);
+        checks.push(...urlContentCheck);
+      }
+      log.info(`Checks for URL ${url}: ${JSON.stringify(checks)}`);
+      return { url, checks };
+    });
+
+    const auditResultsArray = await Promise.all(auditPromises);
```

Comment:
This may result in making 200 calls to fetch the html of the page at the same time.
Also, it one rejects, then the whole is rejected, so you could use Promise.allSettled instead.